core/src/main/java/org/opensearch/sql/calcite/utils/DynamicColumnProcessor.java

```diff
@@ -0,0 +1,239 @@
+  }
+
+  private static List<Column> getColumnsExcludingDynamic(Schema originalSchema) {
+    return originalSchema.getColumns().stream()
+        .filter(c -> !DYNAMIC_COLUMNS_FIELD.equals(c.getName()))
+        .collect(Collectors.toList());
+  }
+
+  private static List<ExprValue> expandResultRowsWithDynamicColumns(
+      List<ExprValue> originalResults, Set<String> dynamicColumnNames) {
+    List<ExprValue> expandedResults = new ArrayList<>();
+
+    for (ExprValue row : originalResults) {
+      if (row instanceof ExprTupleValue) {
+        Map<String, ExprValue> originalRow = row.tupleValue();
+        Map<String, ExprValue> expandedRow = copyExceptDynamic(originalRow);
+
+        Map<String, ExprValue> dynamicColumns =
+            extractMapValue(originalRow.get(DYNAMIC_COLUMNS_FIELD));
+
+        // This will overwrite existing columns if there's a name conflict
+        for (String columnName : dynamicColumnNames) {
+          ExprValue columnValue = dynamicColumns.get(columnName);
+          if (columnValue != null) {
+            expandedRow.put(columnName, columnValue);
+          } else {
+            expandedRow.put(columnName, ExprValueUtils.nullValue());
+          }
+        }
+
+        expandedResults.add(ExprTupleValue.fromExprValueMap(expandedRow));
+      } else {
+        // Non-tuple rows are passed through unchanged
+        expandedResults.add(row);
+      }
+    }
+
+    return expandedResults;
+  }
+
+  private static Map<String, ExprValue> copyExceptDynamic(Map<String, ExprValue> originalRow) {
+    Map<String, ExprValue> copiedRow = new LinkedHashMap<>();
+    for (Map.Entry<String, ExprValue> entry : originalRow.entrySet()) {
+      if (!DYNAMIC_COLUMNS_FIELD.equals(entry.getKey())) {
+        copiedRow.put(entry.getKey(), entry.getValue());
+      }
+    }
+    return copiedRow;
+  }
+
+  private static Map<String, ExprValue> extractMapValue(ExprValue mapValue) {
+    if (mapValue == null || mapValue.isNull() || mapValue.isMissing()) {
+      return Map.of();
+    }
+
+    if (mapValue instanceof ExprTupleValue) {
+      return mapValue.tupleValue();
+    }
+
+    throw new IllegalStateException(
+        "Expected MAP type for dynamic columns field, got: " + mapValue.type());
+  }
+
+  /**
+   * Checks if a field should be resolved as a dynamic column access. This happens when: 1. The
+   * field is not found in the current schema 2. Dynamic columns are available (marked by the
+   * context) 3. We're not in a coalesce function (which has special null handling)
+   */
+  public static boolean tryResolveDynamicField(
+      String fieldName, List<String> currentFields, CalcitePlanContext context) {
+    // Don't resolve dynamic fields in coalesce function (it has special null handling)
+    if (context.isInCoalesceFunction()) {
+      return false;
+    }
+
+    if (currentFields.contains(fieldName)) {
+      return false;
+    }
```

Comment:
current solution resolved field in schema first, if not exist, resolve in _dynamic_col MAP[], which have have 2 behaviours,
* enforce-schema if _dynamic_col not exist, 
* scheam-less if _dynamic_col exist.